### PR TITLE
Feat: Search with nearby cities

### PR DESCRIPTION
**Pull request by timwangmusic**
## Description
Implement the search with nearby cities feature.

## Solution
* Concurrently search for travel plans in nearby cities of the primary city.
* Google Maps API call reduction measures:
- [x]  Limit number of nearby cities within a certain threshold.
- [x]  Prioritize results from the primary city. Only include results from nearby cities when results from the primary city is not sufficient.
- [x]   Fall back to primary city results if nearby cities results have error.

## Testing
- [x] Integration testing on Heroku staging
- [x] Added new unit tests

## Checks
- [x] Have you removed commented code?
- [x] Have you used gofmt to format your code?

### CODE REVIEW

File: planner/planner.go [hunk 1]
```diff
@@ -280,8 +283,52 @@ func (p *MyPlanner) cityStatsHandler(context *gin.Context) {
 	context.JSON(http.StatusOK, view)
 }
 
-func (p *MyPlanner) Planning(ctx context.Context, planningRequest *PlanningReq, user string) (resp PlanningResponse) {
-	planningResponse := p.Solver.Solve(ctx, planningRequest)
+func (p *MyPlanner) Planning(ctx context.Context, planningRequest PlanningReq, user string, withNearbyCities bool) (resp PlanningResponse) {
+	logger := iowrappers.Logger
+	var planningResponse *PlanningResp
```
Comment (rwangsc18) [inline planner/planner.go:288]: This looks fine to me. 
For future improvement, we can also consider adding a common interface with a Solve() method. We can wrap the new multi-city solver logic into struct receiver. In Planning(), we can instantiate a solver instance based on the user input.

File: planner/planner.go [hunk 2]
```diff
@@ -280,8 +283,52 @@ func (p *MyPlanner) cityStatsHandler(context *gin.Context) {
...
+					AdminAreaLevelOne: planningRequest.Location.AdminAreaLevelOne,
+					Country:           planningRequest.Location.Country,
+				},
+				Radius: float64(planningRequest.SearchRadius / 1000),
+				Filter: go_geonames.CityWithPopulationGreaterThan15000,
+			})
+		if err != nil {
+			return PlanningResponse{Err: err}
+		}
+		locations := MapSlice[iowrappers.City, POI.Location](nearbyCityResponse.Cities, toLocation)
```
Comment (rwangsc18) [inline planner/planner.go:317]: How many nearby cities do we usually get? To many nearby cities, can increase the load of the server if the traffic gets big. We can consider adding a cap to the number of parallel search requests.

**Review by rwangsc18 — APPROVED** [review]
Great work! The results also look good on the phone. Left a few comments that you may consider as future improvements.

File: planner/planner.go (hunk 1 above)
Comment (timwangmusic) [reply planner/planner.go:288]: I'll add a boolean field `withNearbyCities` in the `PlanningReq` struct.

File: planner/planner.go (hunk 2 above)
Comment (timwangmusic) [reply planner/planner.go:317]: Good point. Actually in design doc we proposed a limit. Any number in mind?

Comment (rwangsc18) [reply planner/planner.go:317]: We can go with 10 first, and set this cap as a configurable parameter and adjust later. Do you know if Heroku automatically increase the number of servers we are using if the computation load increases?

Comment (timwangmusic) [reply planner/planner.go:317]: I don't think Heroku does that. If we deploy on kubernetes it will do that.

File: planner/planner.go [hunk 3]
```diff
@@ -280,67 +291,130 @@ func (p *MyPlanner) cityStatsHandler(context *gin.Context) {
...
-	if planningResponse.Err != nil {
-		resp.Err = planningResponse.Err
-		resp.StatusCode = planningResponse.ErrorCode
-		return
+	// prioritize planning results from the primary location
+	if resp.Err == nil && len(resp.TravelPlans) >= planningRequest.NumPlans {
+		return resp
+	}
+
+	var lat, lng float64
```
Comment (rwangsc18) [inline planner/planner.go:308]: nit: combine declaration and assignment into one line?

**Review by rwangsc18 — APPROVED** [review]
LGTM. Great work!

File: planner/planner.go (hunk 3 above)
Comment (timwangmusic) [reply planner/planner.go:308]: Done

**Pull request merged**